### DISCUSSION TITLE
How to connect to hundreds of chat rooms?

### DISCUSSION THREAD

**Discussion by janovak** [Teekeks/pyTwitchAPI#299] [Q&A] [answered]
I've connected to hundreds of chat rooms, but I always run into one issue or another after a few minutes/hours. I have my own rate limiting that is stricter than Twitch's, so I'm joining the rooms quite slowly. However, something always happens to the connection. Sometimes I hit `Fatal error on SSL protocol` and sometimes I get `Cannot write to closing transport`.

I got a lot more of these errors before adding an `asyncio.Lock()` to guard writing to the connection. I don't send "real" messages to the chat rooms, but I needed to add the locking to the joining and leaving logic to prevent all the rooms I join and leave from stepping on each other. Although that made things better, it didn't totally prevent connection issues. I'm not sure if there's a concurrency issue in the library itself but wanted to ask about the intended use of the chat portions of this library before digging into it. 

As detecting and restarting connections that have hundreds of chat rooms associated with them isn't practical due to rate limiting issues, I tried creating a connection for each chat room, but I didn't have to join very many rooms before I got a port related issue when establishing the connection. 

In summary, my questions are:
1. What is the correct way to join hundreds of chat rooms?
2. Assuming it's the first way I tried where one connection has hundreds of rooms, am I supposed to add any other locks? My only interaction with the connection is via `join_room`, `leave_room`, and the message handler I registered to process messages sent by other users. As mentioned previously, `join_room` and `leave_room` already have a lock controlling their execution.

**Comment by Teekeks** [comment, marked as answer]
While I have personally tried to mass join 500 channels at once using join_room without any problems (besides it obviously taking forever due to the join rate limit imposed by twitch), there might be some edge cases that you seem to have run into.

I am however not sure if its worth further exploring this for the IRC implementation of Chat since Twitch is phasing this out (see here: https://discuss.dev.twitch.com/t/giving-broadcasters-control-concurrent-join-limits-for-irc-and-eventsub/54997 ) in favor of chat via EventSub.

While the lib already implements the base EventSub topics required, the chat bot implementation does not support Chat via EventSub yet but it is something I am actively working on for the next update.

**Reply by janovak** [reply]
I wasn't aware they were moving away from IRC. Thanks for the heads up! I'll eagerly await the next version of pyTwitchAPI. Any idea of when that will be?

**Reply by Teekeks** [reply]
No ETA yet, sorry
